internal/pkg/cli/deploy/deploy.go

```diff
@@ -0,0 +1,875 @@
+}
+
+// DeployWorkloadInputs is the input of DeployWorkload.
+type DeployWorkloadInputs struct {
+	BuildRequired  bool
+	ImageDigest    string
+	EnvFileARN     string
+	AddonsURL      string
+	RootUserARN    string
+	AppTags        map[string]string
+	ResourceTags   map[string]string
+	ImageRepoURL   string
+	ForceNewUpdate bool
+
+	Ws                     WorkspaceReader
+	NewInterpolator        func(app, env string) Interpolator
+	Unmarshal              func([]byte) (manifest.WorkloadManifest, error)
+	UploadOpts             *UploadCustomResourcesOpts
+	SNSTopicsLister        SNSTopicsLister
+	ServiceDeployer        ServiceDeployer
+	NewSvcUpdater          func(func(*session.Session) ServiceForceUpdater)
+	NewAppVersionGetter    func(string) (VersionGetter, error)
+	PublicCIDRBlocksGetter PublicCIDRBlocksGetter
+	ServiceForceUpdater    ServiceForceUpdater
+	EndpointGetter         EndpointGetter
+	Spinner                Progress
+	now                    func() time.Time
+}
+
+// DeployWorkloadOutputs is the output of DeployWorkload.
+type DeployWorkloadOutputs struct {
+	RDWSAlias     string
+	Subscriptions []manifest.TopicSubscription
+}
+
+// UploadArtifacts uploads the deployment artifacts (image, addons files, env files).
+func (d *WorkloadDeployer) UploadArtifacts(in *UploadArtifactsInputs) (*UploadArtifactsOutputs, error) {
+	if !in.ShouldUploadArtifacts {
+		return &UploadArtifactsOutputs{}, nil
+	}
+	mft, err := d.manifest(&manifestInput{
+		ws:              in.Ws,
+		newInterpolator: in.NewInterpolator,
+		unmarshal:       in.Unmarshal,
+	})
+	if err != nil {
+		return nil, err
+	}
+	imageOutputs, err := d.uploadContainerImage(&uploadContainerImageInputs{
+		ws:               in.Ws,
+		mft:              mft,
+		imgBuilderPusher: in.ImageBuilderPusher,
+	})
+	if err != nil {
+		return nil, err
+	}
+	s3Artifacts, err := d.uploadArtifactsToS3(&uploadArtifactsToS3Inputs{
+		ws:        in.Ws,
+		mft:       mft,
+		fs:        in.Fs,
+		uploader:  in.Uploader,
+		templater: in.Templater,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &UploadArtifactsOutputs{
+		BuildRequired: imageOutputs.buildRequired,
+		ImageDigest:   imageOutputs.imageDigest,
+		EnvFileARN:    s3Artifacts.envFileARN,
+		AddonsURL:     s3Artifacts.addonsURL,
+	}, nil
+}
+
+// DeployWorkload deploys a workload using CloudFormation.
+func (d *WorkloadDeployer) DeployWorkload(in *DeployWorkloadInputs) (*DeployWorkloadOutputs, error) {
+	stackConfigoutputs, err := d.stackConfiguration(in)
+	if err != nil {
+		return nil, err
+	}
+	cmdRunAt := in.now()
+	if err := in.ServiceDeployer.DeployService(os.Stderr, stackConfigoutputs.conf, d.S3Bucket, awscloudformation.WithRoleARN(d.Env.ExecutionRoleARN)); err != nil {
+		var errEmptyCS *awscloudformation.ErrChangeSetEmpty
+		if !errors.As(err, &errEmptyCS) {
+			return nil, fmt.Errorf("deploy service: %w", err)
+		}
+		if !in.ForceNewUpdate {
+			log.Warningln("Set --force to force an update for the service.")
```

Comment:
should we have a `logger` interface or something part of `WorkloadDeployer`?

Reply:
I thought about it but it might be ok since it is within the `cli` package so that we don't output logs everywhere. Why do you want to add a `logger` interface? Like what functionality you think we should have here?